src/main/java/ch/uzh/ifi/hase/soprafs21/security/JwtTokenAuthenticationFilter.java

```diff
@@ -0,0 +1,69 @@
+package ch.uzh.ifi.hase.soprafs21.security;
+
+import ch.uzh.ifi.hase.soprafs21.security.config.SecurityConstants;
+import ch.uzh.ifi.hase.soprafs21.service.JwtTokenUtil;
+import ch.uzh.ifi.hase.soprafs21.service.UserService;
+import io.jsonwebtoken.Claims;
+import org.springframework.security.authentication.UsernamePasswordAuthenticationToken;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.web.authentication.WebAuthenticationDetailsSource;
+import org.springframework.web.filter.OncePerRequestFilter;
+
+import javax.servlet.Filter;
+import javax.servlet.FilterChain;
+import javax.servlet.ServletException;
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+import java.io.IOException;
+
+public class JwtTokenAuthenticationFilter extends OncePerRequestFilter {
+
+    private JwtTokenUtil jwtTokenUtil;
+    private UserService userService;
+
+    public JwtTokenAuthenticationFilter(
+            JwtTokenUtil jwtTokenUtil,
+            UserService userService) {
+
+        this.jwtTokenUtil = jwtTokenUtil;
+        this.userService = userService;
+    }
+
+    @Override
+    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) throws ServletException, IOException {
+        String header = request.getHeader(SecurityConstants.HEADER_STRING);
+
+        if(header == null || !header.startsWith(SecurityConstants.TOKEN_PREFIX)) {
+            response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
+            return;
+        }
+
+        String token = header.replace(SecurityConstants.TOKEN_PREFIX, "");
+
+        if(jwtTokenUtil.validateToken(token, SecurityConstants.SECRET)) {
+            Claims claims = jwtTokenUtil.getClaimsFromJWT(token, SecurityConstants.SECRET);
+            String emailId = claims.getSubject();
+
+            UsernamePasswordAuthenticationToken auth =
+                    userService.getUserByEmail(emailId)
+                            .map(userDetails -> {
+                                UsernamePasswordAuthenticationToken authentication =
+                                        new UsernamePasswordAuthenticationToken(
+                                                emailId, null, null);
```

Comment:
Only the email is not sufficient as SecurityContext, since most of the authorization checks will require at least the user.id as well

Reply:
Could you elaborate on which authorization checks? Because currently, we are authorizing against access token which contains email id of user (as a subject which is a string), so checks performed on that. And further we can always fetch/validate user id against email id.

Reply:
What you just described is authentication of user. I mean the authorization - when we check if the user is authorized to perform some action/request. Example: User A and User B are both authenticated/have cookies and user B sends a request to update the profile of user A. We have to check if he is authorized to do it.

Reply:
But this should be done at individual rest endpoint, whenever it is required like an update. Because the filter is authenticating detail from a header and for cases like you mentioned, data from request body/payload is required which is different for each case

Reply:
Yes, that's what we will do, but an email only in the Security context is not sufficient. See the authorization checks I do in UserService: isRequesterAndAuthenticatedUserTheSame().

Reply:
Exactly, we will call this method from the required API to authorize the request.

Reply:
I am expecting to have the authenticated user present in the SecurityContext. Currently, you set only email.

Reply:
In the below method, you are only using email from SecurityContext
![image](https://user-images.githubusercontent.com/29889416/115126251-f0ae5100-9fcd-11eb-9c05-db850331c0d8.png)
If you can justify why the entire user object is required, we can add the user object in context, but I do not see it as a blocking point for merging this PR.

Reply:
@nehasingh1491 We often need to check if the user is the owner of some resource they are trying to modify to authorize their actions, for which it is much cleaner to have access to all their attributes (at least the ID) than having to retrieve them by email from the DB just to then check their ID.

@KseniaBel we can add this after merging..

Reply:
ok I will add user object in separate task